src/H5Dmpio.c

```diff
@@ -2954,464 +3400,1539 @@ H5D__chunk_redistribute_shared_chunks(const H5D_io_info_t *io_info, const H5D_ty
+                                                  &send_type_derived) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL,
+                        "can't create derived datatypes for chunk re-insertion info")
+
+        /*
+         * Gather information to all ranks for a collective re-insertion
+         * of the modified chunks into the chunk index
+         */
+        H5_CHECK_OVERFLOW(chunk_list_num_entries, size_t, int);
+        if (H5_mpio_gatherv_alloc_simple(chunk_list, (int)chunk_list_num_entries, send_type, recv_type, TRUE,
+                                         0, io_info->comm, mpi_rank, mpi_size, &gathered_array,
+                                         &collective_num_entries) < 0)
+            HGOTO_ERROR(H5E_DATASET, H5E_CANTGATHER, FAIL,
+                        "can't gather chunk index re-insertion info to/from ranks")
+
+        for (i = 0; i < collective_num_entries; i++) {
+            H5D_filtered_collective_io_insert_info_t *coll_entry =
+                &((H5D_filtered_collective_io_insert_info_t *)gathered_array)[i];
+
+            /*
+             * We only need to reinsert this chunk if we had to actually
+             * allocate or reallocate space in the file for it
+             */
+            if (coll_entry->index_info.need_insert) {
+                chunk_ud.chunk_block   = coll_entry->chunk_block;
+                chunk_ud.chunk_idx     = coll_entry->index_info.chunk_idx;
+                chunk_ud.filter_mask   = coll_entry->index_info.filter_mask;
+                chunk_ud.common.scaled = scaled_coords;
+
+                /* Calculate scaled coordinates for the chunk */
+                H5VM_array_calc_pre(chunk_ud.chunk_idx, io_info->dset->shared->ndims,
+                                    io_info->dset->shared->layout.u.chunk.down_chunks, scaled_coords);
+                scaled_coords[io_info->dset->shared->ndims] = 0;
+
+                if ((idx_info->storage->ops->insert)(idx_info, &chunk_ud, io_info->dset) < 0)
+                    HGOTO_ERROR(H5E_DATASET, H5E_CANTINSERT, FAIL,
+                                "unable to insert chunk address into index")
+            }
+        }
+    }
+
+done:
+    if (gathered_array)
+        H5MM_free(gathered_array);
+    if (send_type_derived) {
+        if (MPI_SUCCESS != (mpi_code = MPI_Type_free(&send_type)))
+            HMPI_DONE_ERROR(FAIL, "MPI_Type_free failed", mpi_code)
+    }
+    if (recv_type_derived) {
+        if (MPI_SUCCESS != (mpi_code = MPI_Type_free(&recv_type)))
+            HMPI_DONE_ERROR(FAIL, "MPI_Type_free failed", mpi_code)
+    }
+
+#ifdef H5Dmpio_DEBUG
+    H5D_MPIO_TIME_STOP(mpi_rank);
+    H5D_MPIO_TRACE_EXIT(mpi_rank);
+#endif
+
+    FUNC_LEAVE_NOAPI(ret_value)
+} /* end H5D__mpio_collective_filtered_chunk_reinsert() */
+
+/*-------------------------------------------------------------------------
+ * Function:    H5D__mpio_get_chunk_redistribute_info_types
```

Comment:
The 3 functions below here create different MPI derived datatypes to extract certain portions of information from the overall per-chunk H5D_filtered_collective_io_info_t structure. Usually, a particular operation (shared chunk redistribution, chunk reallocation, chunk reinsertion) only needs a few fields out of that structure and this information is gathered to all ranks, so sending just the few fields necessary can drastically save on memory usage at the expense of a bit of MPI overhead.